### PR TITLE
Changes pg13 as main image and moves pg14 into directory

### PR DESCRIPTION
**Pull request by gurkanindibay**

### CODE REVIEW

File: .github/workflows/publish_docker_images_cron.yml [hunk 1]
```diff
@@ -20,14 +20,14 @@ jobs:
...
-          - postgres_13
+          - postgres_14
           - nightly
     steps:
       - name: Checkout repository
         uses: actions/checkout@v2
 
       - name: Clone tools branch
-        run: git clone -b v0.8.3 --depth=1  https://github.com/citusdata/tools.git tools
+        run: git clone -b docker-13-as-main --depth=1  https://github.com/citusdata/tools.git tools
```
Comment (gurkanindibay) [inline .github/workflows/publish_docker_images_cron.yml:30]: after merging tools PR I will change it

Comment (onurctirtir) [reply .github/workflows/publish_docker_images_cron.yml:30]: I think we can still use v0.8.3 here ?

**Review by onurctirtir — APPROVED** [review]

File: .github/workflows/publish_docker_images_on_manual.yml [hunk 2]
```diff
@@ -26,7 +26,7 @@ jobs:
         uses: actions/checkout@v2
 
       - name: Clone tools branch
-        run: git clone -b v0.8.3 --depth=1  https://github.com/citusdata/tools.git tools
+        run: git clone -b docker-13-as-main --depth=1  https://github.com/citusdata/tools.git tools
```
Comment (hanefi) [inline .github/workflows/publish_docker_images_on_manual.yml:29]: needs update

File: .github/workflows/publish_docker_images_on_push.yml [hunk 3]
```diff
@@ -20,14 +20,14 @@ jobs:
...
-          - postgres_13
+          - postgres_14
           - nightly
     steps:
       - name: Checkout repository
         uses: actions/checkout@v2
 
       - name: Clone tools branch
-        run: git clone -b v0.8.3 --depth=1  https://github.com/citusdata/tools.git tools
+        run: git clone -b docker-13-as-main --depth=1  https://github.com/citusdata/tools.git tools
```
Comment (hanefi) [inline .github/workflows/publish_docker_images_on_push.yml:30]: needs update

File: .github/workflows/update_version.yml [hunk 4]
```diff
@@ -30,7 +30,7 @@ jobs:
         run: sudo apt-get update && sudo apt install libcurl4-openssl-dev libssl-dev
 
       - name: Clone tools branch
-        run: git clone -b v0.8.3 --depth=1  https://github.com/citusdata/tools.git tools
+        run: git clone -b docker-13-as-main --depth=1  https://github.com/citusdata/tools.git tools
```
Comment (hanefi) [inline .github/workflows/update_version.yml:33]: needs update

**Review by hanefi — APPROVED** [review]
needs some updates after https://github.com/citusdata/tools/pull/187/files is merged and we create a new tag.

File: .github/workflows/publish_docker_images_cron.yml (hunk 1 above)
Comment (hanefi) [reply .github/workflows/publish_docker_images_cron.yml:30]: > I think we can still use v0.8.3 here ?

I do not think so. Tools repo decides on which tag will be associated with which Dockerfile, and we updated that logic after v0.8.3

File: Dockerfile [hunk 5]
```diff
@@ -12,22 +12,12 @@ LABEL maintainer="Citus Data https://citusdata.com" \
...
 # install Citus
 RUN apt-get update \
     && apt-get install -y --no-install-recommends \
        ca-certificates \
-       curl \
-    && curl -s https://install.citusdata.com/community/deb.sh |  bash \
-    &&  apt-get install -y postgresql-$PG_MAJOR-citus-10.2=$CITUS_VERSION \
-    && apt-get purge -y --auto-remove curl \
-    && rm -rf /var/lib/apt/lists/*
+    && curl -s https://install.citusdata.com/community/deb.sh | bash
```
Comment (hanefi) [inline Dockerfile:20]:
```suggestion
    curl \
    && curl -s https://install.citusdata.com/community/deb.sh | bash
    &&  apt-get install -y postgresql-$PG_MAJOR-citus-10.2=$CITUS_VERSION \
    && apt-get purge -y --auto-remove curl \
    && rm -rf /var/lib/apt/lists/*
```

You are missing some steps in here

File: postgres-14/Dockerfile [hunk 6]
```diff
@@ -1,6 +1,6 @@
 # This file is auto generated from it's template,
 # see citusdata/tools/packaging_automation/templates/docker/latest/latest.tmpl.dockerfile.
-FROM postgres:13.4
+FROM postgres:14beta3
```
Comment (hanefi) [inline postgres-14/Dockerfile:3]:
```suggestion
FROM postgres:14rc1
```

File: postgres-14/Dockerfile [hunk 7]
```diff
@@ -12,16 +12,20 @@ LABEL maintainer="Citus Data https://citusdata.com" \
       org.label-schema.schema-version="1.0"
 
 ENV CITUS_VERSION ${VERSION}.citus-1
+#install postgresql
+RUN apt-get update && apt-get install -y lsb-release apt-utils wget curl gnupg2 && apt-get clean all \
+    && wget --quiet -O - https://www.postgresql.org/media/keys/ACCC4CF8.asc | apt-key add - \
+    && sh -c 'echo "deb http://apt.postgresql.org/pub/repos/apt $(lsb_release -cs)-pgdg main 14" > /etc/apt/sources.list.d/pgdg.list' \
+    && apt-get purge -y --auto-remove curl \
+    && rm -rf /var/lib/apt/lists/*
```
Comment (hanefi) [inline postgres-14/Dockerfile:20]: I did not try to build the following on my local. Please review it before applying it.

I wanted to improve readability and remove unnecessary commands.

```suggestion
# install postgresql 14
RUN apt-get update \
    && apt-get install -y --no-install-recommends \
       apt-utils \
       gnupg2 \
       lsb-release \
       wget \
    && wget --quiet -O - https://www.postgresql.org/media/keys/ACCC4CF8.asc | apt-key add - \
    && echo "deb http://apt.postgresql.org/pub/repos/apt $(lsb_release -cs)-pgdg main 14" > /etc/apt/sources.list.d/pgdg.list \
    && rm -rf /var/lib/apt/lists/*
```

File: postgres-14/Dockerfile [hunk 8]
```diff
@@ -12,16 +12,20 @@ LABEL maintainer="Citus Data https://citusdata.com" \
...
 RUN apt-get update \
     && apt-get install -y --no-install-recommends \
        ca-certificates \
        curl \
-    && curl -s https://install.citusdata.com/community/deb.sh | bash \
-    && apt-get install -y postgresql-$PG_MAJOR-citus-10.2=$CITUS_VERSION \
-                          postgresql-$PG_MAJOR-hll=2.15.citus-1 \
-                          postgresql-$PG_MAJOR-topn=2.3.1 \
+    && curl -s https://install.citusdata.com/community/deb.sh |  bash \
+    &&  apt-get install -y postgresql-$PG_MAJOR-citus-10.2=$CITUS_VERSION \
```
Comment (hanefi) [inline postgres-14/Dockerfile:28]: Please add topn and hll here as well

**Review by hanefi — CHANGES_REQUESTED** [review]

File: postgres-14/Dockerfile (hunk 6 above)
Comment (gurkanindibay) [reply postgres-14/Dockerfile:3]: 14rc1 is debian 11 based as well so rollbacked into beta3

File: .github/workflows/publish_docker_images_cron.yml [hunk 9]
```diff
@@ -20,17 +20,17 @@ jobs:
...
-          - postgres_13
+          - postgres_14
           - nightly
     steps:
       - name: Checkout repository
         uses: actions/checkout@v2
 
       - name: Clone tools branch
-        run: git clone -b v0.8.3 --depth=1  https://github.com/citusdata/tools.git tools
+        run: git clone -b v0.8.4 --depth=1  https://github.com/citusdata/tools.git tools
```
Comment (hanefi) [inline .github/workflows/publish_docker_images_cron.yml:30]:
```suggestion
        run: git clone -b v0.8.4 --depth=1 https://github.com/citusdata/tools.git tools
```

File: .github/workflows/update_version.yml [hunk 10]
```diff
@@ -27,10 +27,10 @@ jobs:
         uses: actions/checkout@v2
 
       - name: Install package dependencies
-        run: sudo apt-get update && sudo apt install libcurl4-openssl-dev libssl-dev
+        run: sudo apt-get update && sudo apt install libcurl4-openssl-dev libssl-dev python3-testresources
 
       - name: Clone tools branch
-        run: git clone -b v0.8.3 --depth=1  https://github.com/citusdata/tools.git tools
+        run: git clone -b v0.8.4 --depth=1  https://github.com/citusdata/tools.git tools
```
Comment (hanefi) [inline .github/workflows/update_version.yml:33]:
```suggestion
        run: git clone -b v0.8.4 --depth=1 https://github.com/citusdata/tools.git tools
```

File: nightly/Dockerfile [hunk 11]
```diff
@@ -8,14 +8,16 @@ LABEL maintainer="Citus Data https://citusdata.com" \
       org.label-schema.version="Nightly" \
       org.label-schema.schema-version="1.0"
 
-# switch to Citus nightly
+# Perform pre requisites for Citus Nightly Installation
```
Comment (hanefi) [inline nightly/Dockerfile:11]:
```suggestion
# Perform prerequisites for Citus Nightly Installation
```

File: postgres-14/Dockerfile (hunk 7 above)
Comment (hanefi) [reply postgres-14/Dockerfile:20]: Several questions:
1. Why do you need `sh -c` to execute the commands?
2. Why do you need `curl`?
3. Can you provide one package name per line?

File: postgres-14/Dockerfile (hunk 8 above)
Comment (hanefi) [reply postgres-14/Dockerfile:28]: ping

**Review by hanefi — CHANGES_REQUESTED** [review]

File: postgres-14/Dockerfile [hunk 12]
```diff
@@ -12,16 +12,20 @@ LABEL maintainer="Citus Data https://citusdata.com" \
...
 # install Citus
 RUN apt-get update \
     && apt-get install -y --no-install-recommends \
        ca-certificates \
        curl \
-    && curl -s https://install.citusdata.com/community/deb.sh | bash \
-    && apt-get install -y postgresql-$PG_MAJOR-citus-10.2=$CITUS_VERSION \
-                          postgresql-$PG_MAJOR-hll=2.15.citus-1 \
-                          postgresql-$PG_MAJOR-topn=2.3.1 \
+    && curl -s https://install.citusdata.com/community/deb.sh |  bash \
```
Comment (hanefi) [inline postgres-14/Dockerfile:27]:
```suggestion
    && curl -s https://install.citusdata.com/community/deb.sh | bash \
```

File: .github/workflows/publish_docker_images_on_manual.yml [hunk 13]
```diff
@@ -26,10 +26,10 @@ jobs:
         uses: actions/checkout@v2
 
       - name: Clone tools branch
-        run: git clone -b v0.8.3 --depth=1  https://github.com/citusdata/tools.git tools
+        run: git clone -b v0.8.4 --depth=1  https://github.com/citusdata/tools.git tools
```
Comment (hanefi) [inline .github/workflows/publish_docker_images_on_manual.yml:29]:
```suggestion
        run: git clone -b v0.8.4 --depth=1 https://github.com/citusdata/tools.git tools
```

File: nightly/Dockerfile [hunk 14]
```diff
@@ -8,14 +8,16 @@ LABEL maintainer="Citus Data https://citusdata.com" \
...
 RUN apt-key del 1530DF18 \
     && rm -rf /etc/apt/sources.list.d/citusdata_community.list \
     && apt-get update \
     && apt-get install -y --no-install-recommends \
        ca-certificates \
-       curl \
-    && curl -s https://install.citusdata.com/community-nightlies/deb.sh | bash \
+       curl
+
+# Install Citus nightly
```
Comment (hanefi) [inline nightly/Dockerfile:19]: Why do we have 2 separate `RUN` steps here? Can we merge them into one to reduce total docker image size?

File: nightly/Dockerfile [hunk 15]
```diff
@@ -8,14 +8,16 @@ LABEL maintainer="Citus Data https://citusdata.com" \
...
+RUN curl -s https://install.citusdata.com/community-nightlies/deb.sh | bash \
     && apt-get install -y postgresql-$PG_MAJOR-citus \
     && apt-get purge -y --auto-remove curl \
-    && rm -rf /var/lib/apt/lists/*
+    && rm -rf /var/lib/apt/lists/* \
```
Comment (hanefi) [inline nightly/Dockerfile:23]: Why do you need `\` at the end? I guess this is the last line and it is not needed.

File: postgres-14/Dockerfile [hunk 16]
```diff
@@ -12,16 +12,25 @@ LABEL maintainer="Citus Data https://citusdata.com" \
...
+        wget \
+        gnupg2  \
+    && apt-get clean all \
+    && wget --quiet -O - https://www.postgresql.org/media/keys/ACCC4CF8.asc | apt-key add - \
+    && echo "deb http://apt.postgresql.org/pub/repos/apt $(lsb_release -cs)-pgdg main 14" > /etc/apt/sources.list.d/pgdg.list \
+    && apt-get purge -y --auto-remove curl \
+    && rm -rf /var/lib/apt/lists/*
 
 # install Citus
 RUN apt-get update \
```
Comment (hanefi) [inline postgres-14/Dockerfile:28]: It does not make sense to me that we have 2 separate `RUN` commands that start with `apt-get update`. Can we merge that into one?

**Review by hanefi — CHANGES_REQUESTED** [review]

File: nightly/Dockerfile (hunk 14 above)
Comment (gurkanindibay) [reply nightly/Dockerfile]: Fixed

File: nightly/Dockerfile (hunk 15 above)
Comment (gurkanindibay) [reply nightly/Dockerfile]: Fixed

File: postgres-14/Dockerfile (hunk 16 above)
Comment (gurkanindibay) [reply postgres-14/Dockerfile]: Fixed

File: postgres-14/Dockerfile [hunk 17]
```diff
@@ -13,17 +13,27 @@ LABEL maintainer="Citus Data https://citusdata.com" \
...
-                          postgresql-$PG_MAJOR-hll=2.15.citus-1 \
-                          postgresql-$PG_MAJOR-topn=2.3.1 \
-    && apt-get purge -y --auto-remove curl \
-    && rm -rf /var/lib/apt/lists/*
+    && curl -s https://install.citusdata.com/community/deb.sh |  bash \
+    &&  apt-get install -y postgresql-$PG_MAJOR-citus-10.2=$CITUS_VERSION \
+                          postgresql-$PG_MAJOR-hll=2.16.citus-1 \
+                          postgresql-$PG_MAJOR-topn=2.4.0 \
+    && rm -rf /var/lib/apt/lists/* \
+    && apt-get purge -y --auto-remove curl
```
Comment (hanefi) [inline postgres-14/Dockerfile:35]:
```suggestion
    && apt-get purge -y --auto-remove curl
    && rm -rf /var/lib/apt/lists/* \
```

It is a better idea to delete the local repository after we are done with all other `apt` commands.

**Review by hanefi — CHANGES_REQUESTED** [review]

File: postgres-14/Dockerfile [hunk 18]
```diff
@@ -13,17 +13,27 @@ LABEL maintainer="Citus Data https://citusdata.com" \
 
 ENV CITUS_VERSION ${VERSION}.citus-1
 
-# install Citus
-RUN apt-get update \
+#install postgresql repo and citus
+#install postgresql
```
[the remark this answers is not in the document]
Comment (gurkanindibay) [reply postgres-14/Dockerfile:17]: Removed one of them

File: postgres-14/Dockerfile [hunk 19]
```diff
@@ -13,17 +13,27 @@ LABEL maintainer="Citus Data https://citusdata.com" \
...
+RUN apt-get update  \
+    && apt-get install -y lsb-release \
+                    apt-utils \
+                    wget \
+                    curl \
+                    gnupg2  \
+    && apt-get clean all \
```
[the remark this answers is not in the document]
Comment (gurkanindibay) [reply postgres-14/Dockerfile:24]: Removed apt get clean

File: postgres-14/Dockerfile [hunk 20]
```diff
@@ -13,17 +13,27 @@ LABEL maintainer="Citus Data https://citusdata.com" \
...
-# install Citus
-RUN apt-get update \
+#install postgresql repo and citus
+#install postgresql
+RUN apt-get update  \
+    && apt-get install -y lsb-release \
+                    apt-utils \
+                    wget \
+                    curl \
+                    gnupg2  \
```
[the remark this answers is not in the document]
Comment (gurkanindibay) [reply postgres-14/Dockerfile:23]: The second one is with '--no-install-recommends' option. So it is better to keep them seperate

File: postgres-14/Dockerfile (hunk 17 above)
Comment (gurkanindibay) [reply postgres-14/Dockerfile:35]: Fixed

File: postgres-14/Dockerfile [hunk 21]
```diff
@@ -13,17 +13,27 @@ LABEL maintainer="Citus Data https://citusdata.com" \
...
-                          postgresql-$PG_MAJOR-topn=2.3.1 \
-    && apt-get purge -y --auto-remove curl \
-    && rm -rf /var/lib/apt/lists/*
+    && curl -s https://install.citusdata.com/community/deb.sh |  bash \
+    &&  apt-get install -y postgresql-$PG_MAJOR-citus-10.2=$CITUS_VERSION \
+                          postgresql-$PG_MAJOR-hll=2.16.citus-1 \
+                          postgresql-$PG_MAJOR-topn=2.4.0 \
+    && rm -rf /var/lib/apt/lists/* \
+    && apt-get purge -y --auto-remove curl
+
```
[the remark this answers is not in the document]
Comment (gurkanindibay) [reply postgres-14/Dockerfile:36]: Fixed

File: postgres-14/Dockerfile (hunk 20 above)
[the remark this answers is not in the document]
Comment (hanefi) [reply postgres-14/Dockerfile:22]: We can supply the command to both. It is always nice to have `--no-install-recommends` in automated systems to have a slightly more deterministic behavior. I think we should merge them both.

File: postgres-14/Dockerfile (hunk 12 above)
Comment (hanefi) [reply postgres-14/Dockerfile:28]: You did not apply my suggestion here.

File: postgres-14/Dockerfile (hunk 8 above)
Comment (hanefi) [reply postgres-14/Dockerfile:29]: Please remove the extra space here.

**Review by hanefi — CHANGES_REQUESTED** [review]

File: postgres-14/Dockerfile [hunk 22]
```diff
@@ -13,16 +13,21 @@ LABEL maintainer="Citus Data https://citusdata.com" \
 
 ENV CITUS_VERSION ${VERSION}.citus-1
 
-# install Citus
-RUN apt-get update \
-    && apt-get install -y --no-install-recommends \
-       ca-certificates \
-       curl \
+#install postgresql repo and citus
```
Comment (hanefi) [inline postgres-14/Dockerfile:16]: I do not like the fact that we have different steps in these Dockerfiles. This should not be necessary for several reasons.

1. The base image already has the PGDG repo installed. No reason to have explicit steps here.
2. If we keep introducing new Dockerfiles, we are increasing the technical debt and making it harder to maintain this repo.


Please consider the following `RUN` command if it looks ok with you:

```Dockerfile
# install Citus
RUN apt-get update \
    && apt-get install -y --no-install-recommends \
       ca-certificates \
       curl \
    && curl -s https://install.citusdata.com/community/deb.sh | bash \
    && apt-get install -y postgresql-$PG_MAJOR-citus-10.2=$CITUS_VERSION \
                          postgresql-$PG_MAJOR-hll=2.16.citus-1 \
                          postgresql-$PG_MAJOR-topn=2.4.0 \
    && apt-get purge -y --auto-remove curl \
    && rm -rf /var/lib/apt/lists/*
```

I just copied the same steps from another Dockerfile from this repo. LMK if that's ok with U

**Review by hanefi — CHANGES_REQUESTED** [review]

File: postgres-14/Dockerfile (hunk 20 above)
[the remark this answers is not in the document]
Comment (gurkanindibay) [reply postgres-14/Dockerfile]: I merged it as you suggested

File: postgres-14/Dockerfile (hunk 22 above)
Comment (gurkanindibay) [reply postgres-14/Dockerfile:16]: Other than pg14 it is OK. Postgres repo installation differs from others

Comment (gurkanindibay) [reply postgres-14/Dockerfile:16]: For latest and postgres 12 the block is same as you suggested so no need to change I think

Comment (hanefi) [reply postgres-14/Dockerfile:16]: I know that the repo installation is different on pg14. However the base image already handles that, and we do not need to do provide any extra steps as long as our base image is `postgres:14beta3`

Comment (gurkanindibay) [reply postgres-14/Dockerfile]: Thank you for the heads up. I didn't think that base image could have such definitions now fixed as you suggested

File: docker-compose.yml [hunk 23]
```diff
@@ -5,7 +5,7 @@ version: "3"
 services:
   master:
     container_name: "${COMPOSE_PROJECT_NAME:-citus}_master"
-    image: "citusdata/citus:10.1.2"
+    image: "citusdata/citus:10.2.1"
```
Comment (hanefi) [inline docker-compose.yml:8]: You need to update `services.worker.image` property as well

Comment (hanefi) [reply docker-compose.yml:8]: We do not have the images ready yet. So I do not think that this `docker-compose.yml` will work now

Comment (gurkanindibay) [reply docker-compose.yml:8]: Fixed that one as well

**Review by hanefi — APPROVED** [review]
Please update the PR title and description before merging this one.

**Pull request merged**